src/main/java/build/buildfarm/common/DigestUtil.java

```diff
@@ -46,57 +112,51 @@ public InputStream openStream() throws IOException {
     }
   }
 
-  private static byte[] getSHA1Digest(final Path path) throws IOException {
-    // Naive I/O implementation.  TODO(olaola): optimize!
+  private byte[] getDigestOrFail(Path path)
+      throws IOException {
     return new ByteSource() {
       @Override
       public InputStream openStream() throws IOException {
         return Files.newInputStream(path);
       }
-    }.hash(Hashing.sha1()).asBytes();
+    }.hash(hashFn.getHash()).asBytes();
   }
 
-  public static Digest computeDigest(Path file) throws IOException {
-    long fileSize = Files.size(file);
-    byte[] digest = getSHA1Digest(file);
-    return buildDigest(digest, fileSize);
+  public Digest compute(Path file) throws IOException {
+    byte[] digest = getDigestOrFail(file);
+    return buildDigest(digest, Files.size(file));
+  }
+
+  public Digest build(String hexHash, long size) {
+    if (!hashFn.isValidHexDigest(hexHash)) {
+      throw new NumberFormatException("Invalid hash size");
+    }
+    return buildDigest(hexHash, size);
   }
 
   /**
    * Computes a digest of the given proto message. Currently, we simply rely on message output as
    * bytes, but this implementation relies on the stability of the proto encoding, in particular
    * between different platforms and languages. TODO(olaola): upgrade to a better implementation!
    */
-  public static Digest computeDigest(Message message) {
-    return computeDigest(message.toByteString());
-  }
-
-  /**
-   * A special type of Digest that is used only as a remote action cache key. This is a
-   * separate type in order to prevent accidentally using other Digests as action keys.
-   */
-  public static final class ActionKey {
-    private final Digest digest;
-
-    public Digest getDigest() {
-      return digest;
-    }
-
-    private ActionKey(Digest digest) {
-      this.digest = digest;
-    }
+  public Digest compute(Message message) {
+    return compute(message.toByteString());
   }
 
-  public static ActionKey computeActionKey(Action action) {
-    return new ActionKey(computeDigest(action));
+  public DigestUtil.ActionKey computeActionKey(Action action) {
```

Comment:
Why not just ActionKey?

Reply:
why not indeed... broadening this to actually make the Instance defined in terms of it.